tm4c123x-hal/src/eeprom.rs

```diff
@@ -0,0 +1,390 @@
+//! Code for the EEProm module.
+
+use core::convert::TryInto;
+
+use crate::sysctl::{self};
+use tm4c123x::EEPROM;
+pub use tm4c_hal::eeprom::{Blocks, Busy, EepromAddress, EepromError, Erase, Read, Write};
+
+// Number of EEPROM block on the TM4C123
+const EEPROM_BLOCK_SIZE: usize = 16;
+
+// Number of EEPROM blocks on the TM4C123
+const EEPROM_NUM_BLOCKS: usize = 32;
+
+// Total number of bytes in the EEPROM on the TM4C123
+const EEPROM_END_ADDRESS_BYTES: usize = 2048;
+
+// Total number of words in the EEPROM on the TM4C123
+const EEPROM_END_ADDRESS_WORDS: usize = 512;
+
+// Size of the EEPROM word in bytes
+const BYTES_PER_WORD: usize = 4;
+
+/// Eeprom struct
+pub struct Eeprom {
+    /// Eeprom registers
+    eeprom: EEPROM,
+}
+
+impl Eeprom {
+    /// Configures a new EEPROM struct using the datasheet section 8.2.4.2
+    pub fn new(eeprom: EEPROM, _pc: &sysctl::PowerControl) -> Self {
+        let final_eeprom = Eeprom { eeprom };
+
+        // See Section 8.2.4.2 EEPROM Initialization and Configuration
+        // in the datasheet:
+
+        // 0. Power on the EEPROM peripheral
+        sysctl::control_power(
+            _pc,
+            sysctl::Domain::Eeprom,
+            tm4c_hal::sysctl::RunMode::Run,
+            tm4c_hal::sysctl::PowerState::On,
+        );
+
+        // 1. The datasheet calls for at least a 6 cycle delay before polling
+        // the working register. Need to make sure the loop isn't optimized
+        // out.
+        final_eeprom.delay(20);
+
+        // 2. Poll busy
+        final_eeprom.wait();
+
+        // 3. Read PRETRY and ERETRY
+        if final_eeprom.eeprom.eesupp.read().eretry().bit_is_set() {
+            panic!("Eeprom ERETRY bit set, please investigate or stop using the EEPROM peripheral");
+        }
+
+        if final_eeprom.eeprom.eesupp.read().pretry().bit_is_set() {
+            panic!("Eeprom PRETRY bit set, please investigate or stop using the EEPROM peripheral");
+        }
+
+        // 4. Software reset
+        sysctl::reset(_pc, sysctl::Domain::Eeprom);
+
+        // 5. Another delay
+        final_eeprom.delay(20);
+
+        // 6. Poll busy
+        final_eeprom.wait();
+
+        // 7. Recheck PRETRY and ERETRY
+        if final_eeprom.eeprom.eesupp.read().eretry().bit_is_set() {
+            panic!("Eeprom ERETRY bit set, please investigate or stop using the EEPROM peripheral");
+        }
+
+        if final_eeprom.eeprom.eesupp.read().pretry().bit_is_set() {
+            panic!("Eeprom PRETRY bit set, please investigate or stop using the EEPROM peripheral");
+        }
+
+        // 8. All done
+        final_eeprom
+    }
+
+    /// The EERPOM has multiple instances where small delays are needed. This
+    /// function tries introduce unoptimizable delays
+    pub fn delay(&self, est_clk_cycles: usize) {
+        let mut unoptimizable_delay: u32 = 0;
+
+        for _ in 0..est_clk_cycles {
```

Comment:
There's a cortex-m function which will do this.

https://docs.rs/cortex-m/latest/cortex_m/asm/fn.delay.html